src/util/seal.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * seal.c
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <kmyth/kmyth.h>
+#include <kmyth/file_io.h>
+
+#include "pelz_log.h"
+#include "seal.h"
+
+#include "pelz_enclave.h"
+#include "sgx_seal_unseal_impl.h"
+
+#define ENCLAVE_PATH "sgx/pelz_enclave.signed.so"
+
+int seal(char *filename, char **outpath, size_t outpath_size, bool tpm)
+{
+  uint8_t *data = NULL;
+  size_t data_len = 0;
+  uint8_t *sgx_seal = NULL;
+  size_t sgx_seal_len = 0;
+  uint8_t *tpm_seal = NULL;
+  size_t tpm_seal_len = 0;
+
+  pelz_log(LOG_DEBUG, "Seal function");
+  //Validating filename and data from file
+  if (read_validate(filename, &data, &data_len))
+  {
+    return 1;          
+  }
+
+  //SGX sealing of data in nkl format
+  if (seal_nkl(data, data_len, &sgx_seal, &sgx_seal_len))
+  {
+    free(data);
+    return 1;    
+  }
+  free(data);
+
+  //Checking if TPM seal is requested
+  if (tpm)
+  {
+    //TPM sealing of data in ski format
+    if (seal_ski(sgx_seal, sgx_seal_len, &tpm_seal, &tpm_seal_len))
+    {
+      free(sgx_seal);
+      return 1;
+    }
+    free(sgx_seal);
+  }
+
+  //Checking and/or setting output path
+  if (outpath_validate(filename, outpath, outpath_size, tpm))
+  {
+    return 1;
+  }
+
+  //Write bytes to file based on outpath
+  if (tpm)
+  {
+    if (write_bytes_to_file(*outpath, tpm_seal, tpm_seal_len))
+    {
+      pelz_log(LOG_ERR, "error writing data to .ski file ... exiting");
+      free(tpm_seal);
+      if (*outpath != NULL && outpath_size == 0)
+      {
+        free(*outpath);
+      }
+      return 1;
+    }
+    free(tpm_seal);
+  }
+  else
+  {
+    if (write_bytes_to_file(*outpath, sgx_seal, sgx_seal_len))
+    {
+      pelz_log(LOG_ERR, "error writing data to .nkl file ... exiting");
+      free(sgx_seal);
+      if (*outpath != NULL && outpath_size == 0)
+      {
+        free(*outpath);
+      }
+      return 1;
+    }
+    free(sgx_seal);
+  }
+  return 0;
+}
+
+int seal_ski(uint8_t * data, size_t data_len, uint8_t ** data_out, size_t * data_out_len)
+{
+  char *authString = NULL;
+  size_t auth_string_len = 0;
+  const char *ownerAuthPasswd = "";
+  size_t oa_passwd_len = 0;
+  char *cipherString = NULL;
```

Comment:
I don't think pelz supports configuring things like the cipher string or the pcrs. I would just pass NULL in directly.

Reply:
Pelz does not support this configuration but kmyth does.  They are options in the ```tpm2_kmyth_seal``` and ```tpm2_kmyth_unseal``` function.  Pelz allows the admin to set these variables while the default is NULL.  I do not recommend changing it because the current way has more visibility if the variables are changed to take advantage of the full capabilities of the kmyth tpm2 ```seal``` and ``unseal```.